File: c++/linked_list/q206.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
ListNode* reverseList(ListNode* head) {
    if(head==nullptr || head->next==nullptr)
        return head;
    ListNode* prev = head;
    ListNode* curr = head->next;
    head->next = nullptr;

    while(curr != nullptr){
        ListNode* temp = curr->next;
        curr->next = prev;
        prev = curr;
        curr = temp;
    }

    return prev;
}

File: python/linked_list/q206.py
"""
Reverse a singly linked list.
"""
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        #iterative approach
        if head == None or head.next == None:
            return head

        prev = head
        curr = head.next
        head.next = None

        while curr != None:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
        return prev

def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        #recursive approach
        if head == None or head.next == None:
            return head

        ret = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return ret
